src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -383,7 +383,96 @@ protected void InitializeReferences()
+                if (prevlen == 0)
+                {
+                    if (runner.runtextstart == stoppos)
+                    {
+                        skipScan = true;
+                        runner.runmatch = System.Text.RegularExpressions.Match.Empty;
+                    }
+
+                    runner.runtextpos += RightToLeft ? -1 : 1;
+                }
+
+                if (!skipScan)
+                {
+                    runner.Scan(span);
+
+                    // if we got a match, set runmatch to null if quick is true
+                    if (runner.runmatch!._matchcount[0] > 0)
+                    {
+                        if (quick)
+                        {
+                            runner.runmatch = null;
+                        }
+                    }
+                    else
+                    {
+                        if (runner.runtextpos == stoppos)
+                        {
+                            runner.runmatch = System.Text.RegularExpressions.Match.Empty;
+                        }
+                    }
+                }
+
+                Match? m = runner.runmatch;
+                runner.runmatch = null; // Reset runmatch
+                if (m is not null)
+                {
+                    if (!quick)
+                    {
+                        if (m._matchcount[0] > 0)
+                            m.Tidy(runner.runtextpos);
+                        if (m.Text != input)
+                            m.Text = input;
+                    }
+
+                    // If there was a match and the original text was sliced, then add beggining to the index to get the real
+                    // Index of the match.
+                    if (m.Success && beginning != 0)
+                    {
+                        m.AddBeginningToIndex(beginning);
+                    }
+                }
+                return m;
+            }
+            finally
+            {
+                _runner = runner;
+            }
+        }
+
+        internal Match? Run(bool quick, int prevlen, ReadOnlySpan<char> input, int beginning, int length, int startat)
+        {
+            if ((uint)startat > (uint)input.Length)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.startat, ExceptionResource.BeginIndexNotNegative);
+            }
+            if ((uint)length > (uint)input.Length)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.length, ExceptionResource.LengthNotNegative);
+            }
+
+            RegexRunner runner = Interlocked.Exchange(ref _runner, null) ?? CreateRunner();
+            try
+            {
+                runner.InitializeTimeout(internalMatchTimeout);
+                ReadOnlySpan<char> span = input.Slice(beginning, length);
+                runner.InitializeForScan(this, span, 0, span.Length, startat - beginning, quick);
+                runner.InitializeForGo();
```

Comment:
If we need a method to configure state on the runner, we should have it be a single method rather than three :)

Reply:
(Although I see later that at least the timeout initialization is separate out to before a loop, so maybe that makes sense on its own)

Reply:
Agree, as said above, this will be refactored into two:

- InitializeTimeout: which will only be called once particularly when being called by `Run<TState>` which has a loop inside
- InitializeForScan: which will have contents of what today is IntializeForScan and IntializeForGo, which when being called by `Run<TState>` do need to be called for each new iteration of the loop.